src/python/pants/source/filespec.py

```diff
@@ -16,7 +16,7 @@ def glob_to_regex(pattern):
   :returns: A regex string that matches same paths as the input glob does.
   """
   out = ['^']
-  components = pattern.strip('/').replace('.', '[.]').split('/')
+  components = pattern.strip('/').replace('.', '[.]').replace('$', '[$]').split('/')
```

Comment:
I've been operating under the assumption that this code doesn't need to live very long with the v2 engine on the way. But if it does, this should probably switch to `pathspec`, which we already have a dep on.